crates/ruff_linter/src/rules/flake8_annotations/helpers.rs

```diff
@@ -27,3 +36,81 @@ pub(crate) fn is_overload_impl(
         function.name.as_str() == overloaded_name
     }
 }
+
+/// Given a function, guess its return type.
+pub(crate) fn auto_return_type(
+    function: &ast::StmtFunctionDef,
+    target_version: PythonVersion,
+) -> Option<Expr> {
+    // Collect all the `return` statements.
+    let returns = {
+        let mut visitor = ReturnStatementVisitor::default();
+        visitor.visit_body(&function.body);
+        if visitor.is_generator {
+            return None;
+        }
+        visitor.returns
+    };
+
+    // Determine the return type of the first `return` statement.
+    let (return_statement, returns) = returns.split_first()?;
+    let mut return_type = return_statement.value.as_deref().map_or(
+        ResolvedPythonType::Atom(PythonType::None),
+        ResolvedPythonType::from,
+    );
+
+    // Merge the return types of the remaining `return` statements.
+    for return_statement in returns {
+        return_type = return_type.union(return_statement.value.as_deref().map_or(
+            ResolvedPythonType::Atom(PythonType::None),
+            ResolvedPythonType::from,
+        ));
+    }
+
+    match return_type {
+        ResolvedPythonType::Atom(python_type) => type_expr(python_type),
+        ResolvedPythonType::Union(python_types) if target_version >= PythonVersion::Py310 => {
+            // Aggregate all the individual types (e.g., `int`, `float`).
+            let names = python_types
+                .iter()
+                .sorted_unstable()
+                .filter_map(|python_type| type_expr(*python_type))
+                .collect::<Vec<_>>();
+
+            // Wrap in a bitwise union (e.g., `int | float`).
+            Some(union(&names))
```

Comment:
nit: not critical but maybe we could update `union` to take an iterator instead to avoid allocating a vector here